### PR TITLE
Added Pause feature

### PR DESCRIPTION
**Pull request by MarkMeer**
# Description
The changes that were made to the game in this pull request are the following:
- While in a match, the game can be paused by pressing escape, the start button on the PS4 controller or the menu button on the xbox one controller.
- When paused, the match can be resumed using the 'resume' button in the shown menu.
- When paused, the match can be exited by pressing the 'exit game' button in the shown menu.

Test with the changes from this PR [https://github.com/Brick-Studios/BrickEngine/pull/60](url).

## Screenshot(s)
Screenshot 2019-12-07 at 12 41 40

# How can this be tested?
### keyboard
- [ ] Pause the game using a keyboard by pressing escape once.
- [ ] Resume the game using a keyboard by pressing escape in the pause menu.

### controller
- [ ] Pause the game using a controller by pressing start on a PS4 controller or menu on an Xbox one controller once.
- [ ] Resume the game using a controller by pressing start on a PS4 controller or menu on an Xbox one controller when in the pause menu.

### menu
- [ ] The game is resumed when the 'resume' button is pressed.
- [ ] The game can be exited by pressing the 'exit game' button.

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] Any dependent changes have been merged and published in downstream modules
- [X] Included 1 or more screenshots
- [X] Code is const correct

### CODE REVIEW

File: lib/controllers/game_controller.cpp [hunk 1]
```diff
@@ -435,6 +449,23 @@ void GameController::loadEndGameLevel() {
...
+        // Checks whether an intermission to a new level is going on. When this is the case,
+        // the break won't be initiated and the player has to wait for the next level.
+        if(!scene_manager->isSceneActive<IntermissionScene>()){
+            scene_manager->destroyScene(PauseScene::getLayerStatic());
+            scene_manager->createScene<PauseScene>(*entityFactory, *engine, [this](){this->pauseGame();},
+                                                    [this](){this->loadMainMenu();});
+            having_a_break = true;
+        }
+    }
+}
```
Comment (Peter-JanGootzen) [inline lib/controllers/game_controller.cpp]: Isn't `having_a_break` the same as checking whether the PauseScene is currently active?

I don't really like the hard coded way of checking whether the intermission scene is active, you shouldn't have to specify a specific scene here. And you should be able to pause during the intermission, as it really is still in game and you should be able to pause.

**Review by Peter-JanGootzen — CHANGES_REQUESTED** [review]

**Review by JanSchollaert — APPROVED** [review]
epic

**Review by Peter-JanGootzen — APPROVED** [review]
:panda:

File: lib/controllers/game_controller.cpp [hunk 2]
```diff
@@ -203,6 +218,7 @@ void GameController::setupInput() {
     inputMapping[1][InputKeyCode::EKey_pagedown] = PlayerInput::SPEED_DOWN;
     inputMapping[1][InputKeyCode::EKey_pageup] = PlayerInput::SPEED_UP;
     inputMapping[1][InputKeyCode::EKey_home] = PlayerInput::SPEED_RESET;
+    inputMapping[1][InputKeyCode::EKey_escape] = PlayerInput::PAUSE;
```
Comment (Bram-Boris) [inline lib/controllers/game_controller.cpp]: You are not able to pause the game if Player 1 does not exist.

**Review by Bram-Boris — CHANGES_REQUESTED** [review]

File: lib/controllers/game_controller.cpp (hunk 2 above)
Comment (Bram-Boris) [reply lib/controllers/game_controller.cpp]: With the keyboard*

Comment (MarkMeer) [reply lib/controllers/game_controller.cpp]: Shall I change this for the whole line up above as the other function will not be available either?

**Review by Bram-Boris — APPROVED** [review]

**Pull request merged**